source-code-git/fs/src/schema.ts

```diff
@@ -1,7 +1,7 @@
 export type Filesystem = {
 	writeFile: (path: string, content: string) => Promise<void>
-	readFile: (path: string) => Promise<string>
-	readdir: (path: string) => Promise<string[] | undefined>
+	readFile: (path: string) => Promise<string | null>
+	readdir: (path: string) => Promise<string[] | null>
```

Comment:
Is there a reason to return `null` instead of `undefined`? 

TypeScript recommends to never use `null` https://github.com/Microsoft/TypeScript/wiki/Coding-guidelines#null-and-undefined. Douglas Crockford also recommends to never use `null` https://www.youtube.com/watch?v=PSGEjv3Tqo0&t=561s.

Reply:
I find null` and `undefined` to be useful to distinguish between things that have been intentionally defined as having no value, vs things that have not been assigned a value but may be assigned one in the future. Similar to the difference between `404 Not Found` and `410 Gone`. 

That way when I see a `null` value somewhere I can tell that its something that's not supposed to be defined vs something that may have been left undefined due to a runtime error.

Ultimately its more of stylistic choice, and matters less when using typescript, but that's my reasoning at least.

Reply:
Gotcha. I vote to stick with `undefined` for the codebase and avoid JavaScripts two billion dollar mistake of having two nullable values.